### PR TITLE
Group S3 policies by action rather than resource

### PR DESCRIPTION
**Pull request by edwardfoyle**
## Problem

The initial version of storage access grouped IAM policies by S3 prefix. This is problematic because the number of S3 prefixes may be large and cause policy size limit issues. 

Additionally, once we implement deny rules on subpaths, grouping by action will make those policies easier to reason about (at least in my head).

**Issue number, if available:**

## Changes

Group policies by S3 action, which we are only supporting a few of. This will keep the policy size from growing linearly with the number of S3 prefixes.

The updated tests should make it fairly obvious how the policies are being changed. Note that the permissions being granted by the policy is NOT changed by this PR.

There's also a little bit of type refactoring included.

You may notice a few of the test cases where there are multiple actions on a single S3 prefix actually create a larger policy document now. There is a potential optimization we can make to group by both action and prefix when they are the same. This optimization can be made if/when needed in the future.

**Corresponding docs PR, if applicable:**

## Validation

Updated existing unit test coverage.

## Checklist

- [x] If this PR includes a functional change to the runtime behavior of the code, I have added or updated automated test coverage for this change.
- [x] If this PR requires a change to the [Project Architecture README](../PROJECT_ARCHITECTURE.md), I have included that update in this PR.
- [x] If this PR requires a docs update, I have linked to that docs PR above.
- [x] If this PR modifies E2E tests, makes changes to resource provisioning, or makes SDK calls, I have run the PR checks with the `run-e2e` label set.

_By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license._

### CODE REVIEW

File: packages/backend-storage/src/access_builder.ts [hunk 1]
```diff
@@ -1,46 +1,10 @@
 import {
   AuthRoleName,
-  ConstructFactory,
   ConstructFactoryGetInstanceProps,
-  ResourceAccessAcceptor,
   ResourceAccessAcceptorFactory,
   ResourceProvider,
 } from '@aws-amplify/plugin-types';
-
-export type StorageAction = 'read' | 'write' | 'delete';
```
Comment (edwardfoyle) [inline packages/backend-storage/src/access_builder.ts:10]: I moved all of this to the types.ts file

File: packages/backend-storage/src/action_to_resources_map.ts [hunk 2]
```diff
@@ -0,0 +1,72 @@
+import { ResourceAccessAcceptor } from '@aws-amplify/plugin-types';
+import { StorageAction, StoragePrefix } from './types.js';
+
+/**
+ * Internal collaborating class for maintaining the relationship between an acceptor token and the access map
+ */
+export class AcceptorTokenAccessMap {
```
Comment (edwardfoyle) [inline packages/backend-storage/src/action_to_resources_map.ts:7]: These are some internal classes to abstract away all of the "create if not exists" logic of setting elements in these maps

File: packages/backend-storage/src/index.ts [hunk 3]
```diff
@@ -1,9 +1,3 @@
 export * from './factory.js';
 export { StorageResources, AmplifyStorageProps } from './construct.js';
-export {
-  RoleAccessBuilder,
-  StorageAccessDefinition,
-  StorageAccessBuilder,
-  StorageAction,
-} from './access_builder.js';
```
Comment (rtpascual) [inline packages/backend-storage/src/index.ts:8]: Doesn't this result in a change to `API.md`? Might need to update.

File: packages/backend-storage/src/action_to_resources_map.ts [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+      actionMap.set(action, s3Prefix);
+    });
+  };
+
+  getAccessList = () => {
+    const result: AccessEntry[] = [];
+    this.acceptorTokenAccessMap.forEach((value) => {
+      result.push(value);
+    });
+    return result as Readonly<Readonly<AccessEntry>[]>;
```
Comment (0618) [inline packages/backend-storage/src/action_to_resources_map.ts:37]: if we use `Set` for `result`, is it possible to avoid `as`?

**Review by 0618 — APPROVED** [review]

**Review by 0618 — APPROVED** [review]

**Review by 0618 — DISMISSED** [review]

**Comment by 0618** [comment]
Github forced me to approve so many times 😂

File: packages/backend-storage/src/index.ts (hunk 3 above)
Comment (edwardfoyle) [reply packages/backend-storage/src/index.ts:8]: I just moved these to a different file. There's no change to the API

File: packages/backend-storage/src/action_to_resources_map.ts (hunk 4 above)
Comment (edwardfoyle) [reply packages/backend-storage/src/action_to_resources_map.ts:37]: The type assertion is to wrap everything in `Readonly`. We could do `as Readonly<Set<Readonly<AccessEntry>>>` instead but I don't think it really provides any benefit

File: packages/backend-storage/src/index.ts (hunk 3 above)
Comment (rtpascual) [reply packages/backend-storage/src/index.ts:8]: Ah, missed those at the bottom of types

**Review by rtpascual — DISMISSED** [review]

**Comment by rtpascual** [comment]
> Github forced me to approve so many times 😂

@0618 You just approve that much lol

File: packages/backend-storage/src/construct.ts [hunk 5]
```diff
@@ -24,11 +24,12 @@ import {
 import { fileURLToPath } from 'url';
 import { IFunction } from 'aws-cdk-lib/aws-lambda';
 import { S3EventSourceV2 } from 'aws-cdk-lib/aws-lambda-event-sources';
-import { AmplifyStorageTriggerEvent } from './types.js';
 
 // Be very careful editing this value. It is the string that is used to attribute stacks to Amplify Storage in BI metrics
 const storageStackType = 'storage-S3';
 
+export type AmplifyStorageTriggerEvent = 'onDelete' | 'onUpload';
```
Comment (edwardfoyle) [inline packages/backend-storage/src/construct.ts:31]: moved this here to eliminate a circular import between types and construct

File: .changeset/polite-kiwis-brake.md [hunk 6]
```diff
@@ -0,0 +1,5 @@
+---
+'@aws-amplify/backend-storage': patch
```
Comment (sobolk) [inline .changeset/polite-kiwis-brake.md:2]:
```suggestion
'@aws-amplify/backend-storage': minor
```

**Review by sobolk — APPROVED** [review]

**Review by 0618 — APPROVED** [review]

**Review by rtpascual — APPROVED** [review]

**Pull request merged**